contracts/modules/assessment/Assessment.sol

```diff
@@ -262,61 +242,82 @@ contract Assessment is IAssessment, RegistryAware, Multicall {
-    assessment.assessmentData = assessmentData;
-    // finalizedAt, acceptVotes, denyVotes are initialized to 0 by default
+    _assessments[claimId] = assessment;
 
-    emit AssessmentStarted(claimId, assessingGroupId, start, end);
+    emit AssessmentStarted(claimId, assessment.assessingGroupId, assessment.start, assessment.votingEnd);
   }
 
-  function castVote(uint claimId, bool voteSupport, bytes32 ipfsHash) external whenNotPaused(PAUSE_ASSESSMENTS) {
-    // Validate assessor and get assessment data
+  function castVote(uint claimId, bool voteSupport, bytes32 ipfsHash) override external whenNotPaused(PAUSE_ASSESSMENTS) {
     (uint assessorMemberId, Assessment memory assessment) = _validateAssessor(claimId, msg.sender);
 
-    require(!_isVotingClosed(assessment), VotingPeriodEnded());
-
-    Ballot memory previousVote = _ballots[assessorMemberId][claimId];
-    // Undo the vote if already voted
-    if (previousVote.timestamp > 0) {
-      if (previousVote.support) {
-        assessment.acceptVotes--;
-      } else {
-        assessment.denyVotes--;
-      }
-    }
+    require(block.timestamp > assessment.votingEnd, VotingPeriodEnded());
+    require(_ballots[assessorMemberId][claimId].timestamp == 0, AlreadyVoted());
 
     if (voteSupport) {
       assessment.acceptVotes++;
     } else {
       assessment.denyVotes++;
     }
 
-    _tryFinalize(assessment);
     _assessments[claimId] = assessment;
 
     _ballots[assessorMemberId][claimId] = Ballot({
-      support: voteSupport,
-      ipfsHash: ipfsHash,
-      timestamp: uint32(block.timestamp)
+      timestamp: uint32(block.timestamp),
+      support: voteSupport
     });
+    _ballotsMetadata[assessorMemberId][claimId] = ipfsHash;
 
     emit VoteCast(claimId, msg.sender, assessorMemberId, voteSupport, ipfsHash);
   }
 
-  function hasCooldownPassed(uint claimId) external view returns (bool) {
-    return _hasCooldownPassed(_assessments[claimId]);
+  function closeVoting(uint claimId) override external {
+    Assessment memory assessment = _assessments[claimId];
+    require(assessment.start != 0, InvalidClaimId());
+    require(assessment.votingEnd > block.timestamp, VotingAlreadyClosed());
+
+    uint[] memory assessors = getGroupAssessors(assessment.assessingGroupId);
+    uint groupSize = assessors.length;
+    uint totalVotesFromGroup = 0;
+
+    for(uint i = 0; i < groupSize; i++) {
+      if (_ballots[assessors[i]][claimId].timestamp > 0) {
+        totalVotesFromGroup++;
+      } else {
+        // can break early because we need everybody to vote
+        break;
+      }
+    }
+
+    require(totalVotesFromGroup == groupSize, NotEverybodyVoted());
+
+    assessment.votingEnd = block.timestamp.toUint32();
+    _assessments[claimId] = assessment;
+
+    emit AssessmentVotingEndChanged(claimId, assessment.votingEnd);
+  }
+
+  function resetVotingPeriod(uint claimId) override external onlyContracts(C_GOVERNOR) {
+    Assessment memory assessment = _assessments[claimId];
```

Comment:
since we are only reading the start, cooldownPeriod and votingEnd fields once. is it better to use storage instead?